### PR TITLE
Add Application Mode (Install, Maintenance, Normal)

### PR DESCRIPTION
**Pull request by nupplaphil**
As discussed in #4919 , we use $a->module == 'install' as  a replacement for setting Friendica in Install mode.

This PR will add a definitive "install_mode" setting in the App Class.

My changes:
- Added $install_mode in Class App instead
- Extracted the install business logic to a new Class Install
- Replaced all $a->module == 'install' with $a->install_mode

Tested with install wizard and with `bin/console autoinstall`

### CODE REVIEW

File: src/Core/Install.php [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+		}
+	}
+
+	public static function loadDatabase()
+	{
+		$errors = DBStructure::update(false, true, true);
+
+		return $errors;
+	}
+}
```
Comment (MrPetovan) [inline src/Core/Install.php]: Standards: Please add a newline at the end of files.

**Comment by MrPetovan** [comment]
See https://github.com/friendica/friendica/blob/e51a017d920be6a20eb5a3b56f6c13eb907bb790/index.php#L40-L55

**Comment by nupplaphil** [comment]
I thought that the `$install` flag in `index.php` leads to `$a->module = 'install'`
And in `mod/install.php` , I added the $a->install_mode flag (central point)

But I can also set it at the beginning of index.php (but will let the `setInstallMode()` in `mod/install.php` just to be sure)

**Comment by nupplaphil** [comment]
What about merging `$maintenance` and `$install` to `$a->mode`:
- normal ... Normal Mode of Friendica, everything works
- maintenance ... Maintenance mode
- install ... Install mode

And add a `setAppMode()` in the Class `App` ?

We have just to be carefull about `$a->mode` that it only can take three values (ENUM)

**Comment by MrPetovan** [comment]
Sounds good!

**Comment by nupplaphil** [comment]
Hm.. There are multiple options how to create this `$a->mode` enum.

What do you think is best practice for Friendica?

- Adding a [SplEnum](https://secure.php.net/manual/de/class.splenum.php) Class
- Adding a Int-field with checks
- Adding a String-field and with checks
- Adding two boolean flags in a separated class and make a `getMode` function with boolean-operation

I'm not sure ...

**Comment by MrPetovan** [comment]
I'm not okay with the SPL type if we can't ensure they're available on every basic install of PHP, including shared hosting.

We have been doing a lot of int fields with named constants for the value, but please don't add anything to `boot.php`.

**Comment by nupplaphil** [comment]
I finally replaced any maintenance and install flags with the new App-Mode flag "$a->mode"

Now I better understand what you meant with "do not add anything to `boot.php`", because almost every constant definitions are in boot.php and you try to get rid of it ;-)

File: src/App.php [hunk 2]
```diff
@@ -14,6 +14,10 @@
 
 use Exception;
 
+define ('APP_MODE_NORMAL', 0);
+define ('APP_MODE_INSTALL', 1);
+define ('APP_MODE_MAINTENANCE', 2);
```
Comment (MrPetovan) [inline src/App.php]: Please define these constants inside the `App` class.

**Comment by nupplaphil** [comment]
btw
- Tested normal mode
- Tested Install mode through wizard
- Tested Install mode through auto-install
- Tested maintenance mode

File: src/App.php (hunk 2 above)
Comment (nupplaphil) [reply src/App.php]: For sure? 

I saw it at 
https://github.com/friendica/friendica/blob/8a7c9d6666336eddc877e57ec134e585b4edcf84/src/Network/FKOAuthDataStore.php#L19
and thought I do it the same way ;-)

Comment (MrPetovan) [reply src/App.php]: 😬😬😬

This class is part of the "legacy code I don't want to touch because I'm not sure it would keep working if it ever did".

Comment (nupplaphil) [reply src/App.php]: ah, oh... I see ;-)

So I will add it to the class and make some useful methods around it

**Comment by nupplaphil** [comment]
Please don't merge  this PR yet

There are changes in dev repo for `mod/install.php` which I have to merge (smarty3 template check)

**Comment by MrPetovan** [comment]
I was about to tell you, but you beat me to the post, I can't merge when there's a conflict anyway.

File: mod/install.php [hunk 3]
```diff
@@ -51,7 +51,7 @@ function install_post(App $a) {
 			$phpath = notags(trim($_POST['phpath']));
 
 			require_once("include/dba.php");
-			if (!dba::connect($dbhost, $dbuser, $dbpass, $dbdata)) {
+			if (!dba::connect($dbhost, $dbuser, $dbpass, $dbdata, true)) {
```
Comment (MrPetovan) [inline mod/install.php]: You removed the fifth parameter of the `dba::connect` method, this shouldn't have been readded.

File: src/Core/Console/AutomaticInstallation.php [hunk 4]
```diff
@@ -5,9 +5,9 @@
 use Asika\SimpleConsole\Console;
 use dba;
 use Friendica\App;
+use Friendica\Core\Install;
 
 require_once 'mod/install.php';
-require_once 'include/dba.php';
```
Comment (MrPetovan) [inline src/Core/Console/AutomaticInstallation.php]: Please keep the include until we have moved the `dba` class out of `include/`.

File: src/Core/Install.php [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+ * @brief Contains the class with install relevant stuff *
+ */
+class Install extends BaseObject
+{
+	public static function setInstallMode()
+	{
+		self::getApp()->mode = App::MODE_INSTALL;
+	}
+
+	public static function check($phpath = 'php')
```
Comment (MrPetovan) [inline src/Core/Install.php]: Please add A DocBlock comment for this function, including a precise description of the returned data structure.

File: src/Core/Install.php [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+					$v = $v && $c['status'];
+				}
+				return $v;
+			},
+			true);
+
+		return array($checks, $checkspassed);
+	}
+
+	public static function install($urlpath, $dbhost, $dbuser, $dbpass, $dbdata, $phpath, $timezone, $language, $adminmail, $rino = 1)
```
Comment (MrPetovan) [inline src/Core/Install.php]: Please add a DocBlock comment for this function detailing the parameters.

File: src/Core/Install.php [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+		$checks[] = [
+			'title' => $title,
+			'status' => $status,
+			'required' => $required,
+			'help' => $help,
+			'error_msg' => $error_msg,
+		];
+	}
+
+	public static function checkPHP(&$phpath, &$checks)
```
Comment (MrPetovan) [inline src/Core/Install.php]: Please add a DocBlock comment for this function detailing the parameters.

File: src/Core/Install.php [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+				$help .= L10n::t('The command line version of PHP on your system does not have "register_argc_argv" enabled.') . EOL;
+				$help .= L10n::t('This is required for message delivery to work.');
+			}
+			self::addCheck($checks, L10n::t('PHP register_argc_argv'), $passed3, true, $help);
+		}
+
+
+	}
+
+	public static function checkKeys(&$checks)
```
Comment (MrPetovan) [inline src/Core/Install.php]: Please add a DocBlock comment for this function detailing the parameters.

File: src/Core/Install.php [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+	{
+
```
Comment (MrPetovan) [inline src/Core/Install.php]: Extra line feed.

File: src/Core/Install.php [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+		}
+
+		// Get private key
+
+		if (!$res) {
+			$help .= L10n::t('Error: the "openssl_pkey_new" function on this system is not able to generate encryption keys') . EOL;
+			$help .= L10n::t('If running under Windows, please see "http://www.php.net/manual/en/openssl.installation.php".');
+		}
+		self::addCheck($checks, L10n::t('Generate encryption keys'), $res, true, $help);
+
```
Comment (MrPetovan) [inline src/Core/Install.php]: Extra line feed.

File: src/Core/Install.php [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+			list($result) = explode("\n", $result);
+			$help = "";
+			if (!$passed2) {
+				$help .= L10n::t("PHP executable is not the php cli binary \x28could be cgi-fgci version\x29") . EOL;
+				$help .= L10n::t('Found PHP version: ') . "<tt>$result</tt>";
+			}
+			self::addCheck($checks, L10n::t('PHP cli binary'), $passed2, true, $help);
+		}
+
+
```
Comment (MrPetovan) [inline src/Core/Install.php]: Extra line feed.

File: src/Core/Install.php [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+			$passed3 = $result == $str;
+			$help = "";
+			if (!$passed3) {
+				$help .= L10n::t('The command line version of PHP on your system does not have "register_argc_argv" enabled.') . EOL;
+				$help .= L10n::t('This is required for message delivery to work.');
+			}
+			self::addCheck($checks, L10n::t('PHP register_argc_argv'), $passed3, true, $help);
+		}
+
+
```
Comment (MrPetovan) [inline src/Core/Install.php]: Extra line feed.

File: src/Core/Install.php [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+		if (!$res) {
+			$help .= L10n::t('Error: the "openssl_pkey_new" function on this system is not able to generate encryption keys') . EOL;
+			$help .= L10n::t('If running under Windows, please see "http://www.php.net/manual/en/openssl.installation.php".');
+		}
+		self::addCheck($checks, L10n::t('Generate encryption keys'), $res, true, $help);
+
+	}
+
+
+	public static function checkFunctions(&$checks)
```
Comment (MrPetovan) [inline src/Core/Install.php]: Please add a DocBlock comment for this function detailing the parameters.

File: src/Core/Install.php [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+		try {
+			$xml = new DOMDocument();
+		} catch (Exception $e) {
+			$ck_funcs[5]['status'] = false;
+			$ck_funcs[5]['help'] = L10n::t('Error, XML PHP module required but not installed.');
+		}
+	}
+
+
+	public static function checkHtConfig(&$checks)
```
Comment (MrPetovan) [inline src/Core/Install.php]: Please add a DocBlock comment for this function detailing the parameters.

File: src/Core/Install.php [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+			$help .= L10n::t('This is most often a permission setting, as the web server may not be able to write files in your folder - even if you can.') . EOL;
+			$help .= L10n::t('At the end of this procedure, we will give you a text to save in a file named .htconfig.php in your Friendica top folder.') . EOL;
+			$help .= L10n::t('You can alternatively skip this procedure and perform a manual installation. Please see the file "INSTALL.txt" for instructions.') . EOL;
+		}
+
+		self::addCheck($checks, L10n::t('.htconfig.php is writable'), $status, false, $help);
+
+	}
+
+	public static function checkSmarty3(&$checks)
```
Comment (MrPetovan) [inline src/Core/Install.php]: Please add a DocBlock comment for this function detailing the parameters.

File: src/Core/Install.php [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+			$help .= L10n::t('In order to store these compiled templates, the web server needs to have write access to the directory view/smarty3/ under the Friendica top level folder.') . EOL;
+			$help .= L10n::t("Please ensure that the user that your web server runs as \x28e.g. www-data\x29 has write access to this folder.") . EOL;
+			$help .= L10n::t("Note: as a security measure, you should give the web server write access to view/smarty3/ only--not the template files \x28.tpl\x29 that it contains.") . EOL;
+		}
+
+		self::addCheck($checks, L10n::t('view/smarty3 is writable'), $status, true, $help);
+
+	}
+
+	public static function checkHtAccess(&$checks)
```
Comment (MrPetovan) [inline src/Core/Install.php]: Please add a DocBlock comment for this function detailing the parameters.

**Comment by nupplaphil** [comment]
> Thanks for the code standards work, but it seems you reverted one of your own changes.
Do you have some automatic checks for such mistakes or did you find it by manually checking my changes?

File: src/Core/Install.php [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+
+			$status = false;
+			$help = L10n::t('Friendica uses the Smarty3 template engine to render its web views. Smarty3 compiles templates to PHP to speed up rendering.') . EOL;
+			$help .= L10n::t('In order to store these compiled templates, the web server needs to have write access to the directory view/smarty3/ under the Friendica top level folder.') . EOL;
+			$help .= L10n::t("Please ensure that the user that your web server runs as \x28e.g. www-data\x29 has write access to this folder.") . EOL;
+			$help .= L10n::t("Note: as a security measure, you should give the web server write access to view/smarty3/ only--not the template files \x28.tpl\x29 that it contains.") . EOL;
+		}
+
+		self::addCheck($checks, L10n::t('view/smarty3 is writable'), $status, true, $help);
+
```
Comment (MrPetovan) [inline src/Core/Install.php]: Extra line feed.

File: src/Core/Install.php [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+				$error_msg['msg'] = $test['error'];
+			}
+			self::addCheck($checks, L10n::t('Url rewrite is working'), $status, true, $help);
+		} else {
+			// cannot check modrewrite if libcurl is not installed
+			/// @TODO Maybe issue warning here?
+		}
+	}
+
+	public static function checkImagik(&$checks)
```
Comment (MrPetovan) [inline src/Core/Install.php]: Please add a DocBlock comment for this function detailing the parameters.

File: src/Core/Install.php [hunk 19]
```diff
@@ -0,0 +1,340 @@
...
+			self::addCheck($checks, L10n::t('ImageMagick PHP extension is not installed'), $imagick, false, "");
+		} else {
+			self::addCheck($checks, L10n::t('ImageMagick PHP extension is installed'), $imagick, false, "");
+			if ($imagick) {
+				self::addCheck($checks, L10n::t('ImageMagick supports GIF'), $gif, false, "");
+			}
+		}
+	}
+
+	public static function loadDatabase()
```
Comment (MrPetovan) [inline src/Core/Install.php]: Shouldn't this method called `updateDatabaseStructure` instead?

**Comment by nupplaphil** [comment]
hui .. I will document the methods tomorrow, hopefully there are no new changes in `mod/install.php` meanwhile ;-)

**Comment by MrPetovan** [comment]
Take your time, I'll be available tomorrow as well.

**Comment by nupplaphil** [comment]
Oh no ^^ Another merge for `App` is necessary ^^

**Comment by MrPetovan** [comment]
Thank you so much for bearing with me!

**Pull request merged**